XUtils.h

```diff
@@ -19,11 +19,11 @@ in the source distribution for its full text.
 
 void fail(void) ATTR_NORETURN;
 
-void* xMalloc(size_t size);
+void* xMalloc(size_t size) ATTR_ALLOC_SIZE1(1) ATTR_MALLOC;
 
-void* xCalloc(size_t nmemb, size_t size);
+void* xCalloc(size_t nmemb, size_t size) ATTR_ALLOC_SIZE2(1, 2) ATTR_MALLOC;
 
-void* xRealloc(void* ptr, size_t size);
+void* xRealloc(void* ptr, size_t size) ATTR_ALLOC_SIZE1(2);
```

Comment:
Quoting from [GCC documentation](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html) (emphasis mine):

> This tells the compiler that a function is `malloc`-like, i.e., that **the pointer `P` returned by the function cannot alias any other pointer valid when the function returns**, and moreover **no pointers to valid objects occur in any storage addressed by `P`**.
>
> Using this attribute can improve optimization. Compiler predicts that a function with the attribute returns non-null in most cases. Functions like `malloc` and `calloc` have this property because they return a pointer to uninitialized or zeroed-out storage. **However, functions like `realloc` do not have this property, as they can return a pointer to storage containing pointers.**

Maybe also include `returns_nonnull` for the flags we use? As we force `abort` on error, there is no way the function may return `NULL`.

Reply:
Maybe at some point..